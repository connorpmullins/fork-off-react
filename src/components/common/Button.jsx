import React from "react";
import PropTypes from "prop-types";
import { colors, typography, spacing, borderRadius } from "../../styles/theme";

const VARIANTS = {
  primary: {
    backgroundColor: colors.primary.main,
    color: colors.neutral.white,
    hoverBg: colors.primary.dark,
    activeBg: colors.primary.light,
  },
  secondary: {
    backgroundColor: colors.secondary.main,
    color: colors.neutral.white,
    hoverBg: colors.secondary.dark,
    activeBg: colors.secondary.light,
  },
  outline: {
    backgroundColor: "transparent",
    color: colors.primary.main,
    border: `2px solid ${colors.primary.main}`,
    hoverBg: colors.primary.light,
    hoverColor: colors.neutral.white,
  },
  ghost: {
    backgroundColor: "transparent",
    color: colors.primary.main,
    hoverBg: colors.neutral.gray100,
  },
};

const SIZES = {
  sm: {
    padding: `${spacing[2]} ${spacing[3]}`,
    fontSize: typography.fontSize.sm,
    minWidth: "80px",
  },
  md: {
    padding: `${spacing[3]} ${spacing[4]}`,
    fontSize: typography.fontSize.base,
    minWidth: "120px",
  },
  lg: {
    padding: `${spacing[3]} ${spacing[4]}`,
    fontSize: typography.fontSize.base,
    minWidth: "140px",
  },
};

const Button = ({
  variant = "primary",
  size = "md",
  fullWidth = false,
  disabled = false,
  children,
  onClick,
  type = "button",
  className = "",
  tooltip = "",
}) => {
  const variantStyles = VARIANTS[variant];
  const sizeStyles = SIZES[size];

  const tooltipStyles = {
    position: "absolute",
    bottom: "calc(100% + 8px)",
    left: "50%",
    transform: "translateX(-50%)",
    backgroundColor: colors.neutral.gray800,
    color: colors.neutral.white,
    padding: `${spacing[2]} ${spacing[3]}`,
    borderRadius: borderRadius.md,
    fontSize: typography.fontSize.sm,
    whiteSpace: "nowrap",
    opacity: 0,
    visibility: "hidden",
    transition: "all 150ms ease-in-out",
    zIndex: 10,
    "&::after": {
      content: '""',
      position: "absolute",
      top: "100%",
      left: "50%",
      transform: "translateX(-50%)",
      border: "6px solid transparent",
      borderTopColor: colors.neutral.gray800,
    },
  };

  const containerStyles = {
    position: "relative",
    display: fullWidth ? "block" : "inline-block",
    width: fullWidth ? "100%" : "auto",
  };

  const baseStyles = {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    fontFamily: typography.fontFamily.primary,
    fontWeight: typography.fontWeight.medium,
    borderRadius: borderRadius.md,
    transition: "all 150ms ease-in-out",
    cursor: disabled ? "not-allowed" : "pointer",
    opacity: disabled ? 0.6 : 1,
    width: "100%",
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis",
    border: "none",
    outline: "none",
    ...variantStyles,
    ...sizeStyles,
  };

  return (
    <div
      style={containerStyles}
      onMouseEnter={(e) => {
        if (disabled && tooltip) {
          const tooltipEl = e.currentTarget.querySelector("[data-tooltip]");
          if (tooltipEl) {
            tooltipEl.style.opacity = "1";
            tooltipEl.style.visibility = "visible";
          }
        }
      }}
      onMouseLeave={(e) => {
        if (disabled && tooltip) {
          const tooltipEl = e.currentTarget.querySelector("[data-tooltip]");
          if (tooltipEl) {
            tooltipEl.style.opacity = "0";
            tooltipEl.style.visibility = "hidden";
          }
        }
      }}
      data-testid={`${children.toLowerCase().replace(/\s+/g, "-")}-button-container`}
    >
      {disabled && tooltip && (
        <div style={tooltipStyles} data-tooltip>
          {tooltip}
        </div>
      )}
      <button
        type={type}
        onClick={onClick}
        disabled={disabled}
        style={baseStyles}
        className={className}
      >
        {children}
      </button>
    </div>
  );
};

Button.propTypes = {
  variant: PropTypes.oneOf(["primary", "secondary", "outline", "ghost"]),
  size: PropTypes.oneOf(["sm", "md", "lg"]),
  fullWidth: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
  type: PropTypes.oneOf(["button", "submit", "reset"]),
  className: PropTypes.string,
  tooltip: PropTypes.string,
};

export default Button;
