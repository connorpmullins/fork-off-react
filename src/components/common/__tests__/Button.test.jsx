import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Button from "../Button";

describe("Button", () => {
  it("matches snapshot with default props", () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with different variants", () => {
    const { container, rerender } = render(
      <Button variant="primary">Primary</Button>,
    );
    expect(container).toMatchSnapshot();

    rerender(<Button variant="outline">Outline</Button>);
    expect(container).toMatchSnapshot();

    rerender(<Button variant="secondary">Secondary</Button>);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot when disabled with tooltip", () => {
    const { container } = render(
      <Button disabled tooltip="Please fill required fields">
        Disabled Button
      </Button>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders with default props", () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole("button", { name: /click me/i });
    expect(button).toBeInTheDocument();
  });

  it("applies different variants correctly", () => {
    const { rerender } = render(<Button variant="primary">Primary</Button>);
    let button = screen.getByRole("button");
    expect(button).toHaveStyle({ backgroundColor: "#2563EB" }); // primary color

    rerender(<Button variant="outline">Outline</Button>);
    button = screen.getByRole("button");
    expect(button).toHaveStyle({ backgroundColor: "transparent" });
    expect(button).toHaveStyle({ border: "2px solid #2563EB" });
  });

  it("handles disabled state", () => {
    render(<Button disabled>Disabled</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
    expect(button).toHaveStyle({ cursor: "not-allowed" });
  });

  it("shows tooltip when disabled and hovered", async () => {
    render(
      <Button disabled tooltip="Please fill required fields">
        Disabled
      </Button>,
    );

    const container = screen.getByTestId("disabled-button-container");
    const tooltip = screen.getByText("Please fill required fields");

    expect(tooltip).toHaveStyle({ visibility: "hidden", opacity: "0" });

    fireEvent.mouseEnter(container);
    expect(tooltip).toHaveStyle({ visibility: "visible", opacity: "1" });

    fireEvent.mouseLeave(container);
    expect(tooltip).toHaveStyle({ visibility: "hidden", opacity: "0" });
  });

  it("handles click events when not disabled", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me</Button>);

    fireEvent.click(screen.getByRole("button"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("does not call onClick when disabled", () => {
    const handleClick = jest.fn();
    render(
      <Button disabled onClick={handleClick}>
        Click me
      </Button>,
    );

    fireEvent.click(screen.getByRole("button"));
    expect(handleClick).not.toHaveBeenCalled();
  });

  it("applies different sizes correctly", () => {
    const { rerender } = render(<Button size="sm">Small</Button>);
    expect(screen.getByRole("button")).toHaveStyle({ fontSize: "0.875rem" });

    rerender(<Button size="lg">Large</Button>);
    expect(screen.getByRole("button")).toHaveStyle({ minWidth: "140px" });
  });

  it("handles full width styling", () => {
    render(<Button fullWidth>Full Width</Button>);
    expect(screen.getByTestId("full-width-button-container")).toHaveStyle({
      width: "100%",
    });
  });
});
