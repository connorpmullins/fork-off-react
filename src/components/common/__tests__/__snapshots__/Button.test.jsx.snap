// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot when disabled with tooltip 1`] = `
<div>
  <div
    data-testid="disabled-button-button-container"
    style="position: relative; display: inline-block; width: auto;"
  >
    <div
      data-tooltip="true"
      style="position: absolute; bottom: calc(100% + 8px); left: 50%; transform: translateX(-50%); background-color: rgb(30, 41, 59); color: rgb(255, 255, 255); padding: 0.5rem 0.75rem; border-radius: 0.375rem; font-size: 0.875rem; white-space: nowrap; opacity: 0; visibility: hidden; transition: all 150ms ease-in-out; z-index: 10;"
    >
      Please fill required fields
    </div>
    <button
      class=""
      disabled=""
      style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: not-allowed; opacity: 0.6; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(37, 99, 235); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
      type="button"
    >
      Disabled Button
    </button>
  </div>
</div>
`;

exports[`Button matches snapshot with default props 1`] = `
<div>
  <div
    data-testid="click-me-button-container"
    style="position: relative; display: inline-block; width: auto;"
  >
    <button
      class=""
      style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: pointer; opacity: 1; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(37, 99, 235); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
      type="button"
    >
      Click me
    </button>
  </div>
</div>
`;

exports[`Button matches snapshot with different variants 1`] = `
<div>
  <div
    data-testid="primary-button-container"
    style="position: relative; display: inline-block; width: auto;"
  >
    <button
      class=""
      style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: pointer; opacity: 1; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(37, 99, 235); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
      type="button"
    >
      Primary
    </button>
  </div>
</div>
`;

exports[`Button matches snapshot with different variants 2`] = `
<div>
  <div
    data-testid="outline-button-container"
    style="position: relative; display: inline-block; width: auto;"
  >
    <button
      class=""
      style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: pointer; opacity: 1; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: transparent; color: rgb(37, 99, 235); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px; border: 2px solid #2563eb;"
      type="button"
    >
      Outline
    </button>
  </div>
</div>
`;

exports[`Button matches snapshot with different variants 3`] = `
<div>
  <div
    data-testid="secondary-button-container"
    style="position: relative; display: inline-block; width: auto;"
  >
    <button
      class=""
      style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: pointer; opacity: 1; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(16, 185, 129); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
      type="button"
    >
      Secondary
    </button>
  </div>
</div>
`;
